### PR TITLE
feat: added estimate gas fn to common-files

### PR DESCRIPTION
**Pull request by imagobea**
## What does this implement/fix? Explain your changes.

Continues the [task that started in SDK repo](https://github.com/maticnetwork/nightfall-sdk/pull/133). I have moved the `estimateGas` function and its tests to `common-files`.

## Does this close any currently open issues?

No

## What commands can I run to test the change?

`cd common-files && npm ci`, then run the tests with `npm run test`.

## Any other comments?

First experience with `mocha` and `sinon`, please take a look at the stubs and assertions :pray: Thank you!!

### CODE REVIEW

File: common-files/utils/ethereum/gas.mjs [hunk 1]
```diff
@@ -0,0 +1,32 @@
+import logger from '../logger.mjs';
+
+export const TX_GAS_DEFAULT = 4000000;
```
Comment (israelboudoux) [inline common-files/utils/ethereum/gas.mjs:3]: I think some of these constants already exist [here](https://github.com/EYBlockchain/nightfall_3/blob/master/cli/lib/constants.mjs).

**Review by israelboudoux — APPROVED** [review]
LGTM, there is just a simple detail to double-check.

File: common-files/utils/ethereum/gas.mjs (hunk 1 above)
Comment (pawelgrzybek) [reply common-files/utils/ethereum/gas.mjs:3]: @israelboudoux, @imagobea did the right thing. We should not be importing constants from another top-level package. Alternatively we can take from from config. Personally I would keep it as is. Two repetitions doesn't hurt my DRY-mania :)

**Comment by pawelgrzybek** [comment]
Oooo, that is going to be a good opportunity to test if the automated common-files version on merge to master works as expected 🤞

File: cli/lib/nf3.mjs [hunk 2]
```diff
@@ -276,11 +233,6 @@ class Nf3 {
     let signed;
 
     await this.nonceMutex.runExclusive(async () => {
-      // estimate the gasPrice
```
Comment (druiz0992) [inline cli/lib/nf3.mjs:279]: Why did you remove this?

File: cli/lib/nf3.mjs [hunk 3]
```diff
@@ -292,10 +244,12 @@ class Nf3 {
         to: contractAddress,
         data: unsignedTransaction,
         value: fee,
-        gas,
-        gasPrice,
         nonce: this.nonce,
       };
+      // Estimate the gasLimit
+      const gas = await estimateGas(tx, this.web3);
```
Comment (druiz0992) [inline cli/lib/nf3.mjs:250]: Don’t we need gasPrice as well?

Comment (imagobea) [reply cli/lib/nf3.mjs:250]: @druiz0992 

A quick search revealed that gas calculations were outdated due to London Upgrade of August 2021. Quick summary [1]:

**Prior to the London upgrade**

```
Total fee = Gas units (limit) * Gas price per unit
```

**After the London upgrade**

```
Total fee = Gas units * (base fee + priority fee)
```

_Where the base fee is a value set by the protocol and the priority fee is a value set by the user as a tip to the validator._

**What this means re transaction object**

Before we had `gas` and `gasPrice`. Now you can set [2]:

- `maxFeePerGas`: (optional, defaulted to (`2 * block.baseFeePerGas) + maxPriorityFeePerGas`) The maximum fee per gas that the transaction is willing to pay in total

- `maxPriorityFeePerGas`: (optional, defaulted to 2.5 Gwei) The maximum fee per gas to give miners to incentivise them to include the transaction (Priority fee)

**About this PR**

Limits gas calculations to `estimateGas` function via web3 - this is for calculating "gas units", a.k.a. `gas`.
In the future, we can consider using dynamic values for `maxPriorityFeePerGas` and `maxFeePerGas`.

Comment (imagobea) [reply cli/lib/nf3.mjs:250]: **References**

[1] https://ethereum.org/en/developers/docs/gas
[2] https://web3js.readthedocs.io/en/v1.7.3/web3-eth.html#id88

**Review by signorecello — COMMENTED** [review]
LGTM

File: cli/lib/nf3.mjs (hunk 3 above)
Comment (druiz0992) [reply cli/lib/nf3.mjs:261]: i still seem to remember that when we deployed last time to testnet (beginning of november), we had to update gasPrice because otherwise it wouldn't work.
```
WARN [10-26|14:36:39.709] Served eth_call                          reqid=22578            duration="199.712µs"  err="err: max fee per gas less than block base fee: address 0xe327221Fc90f4D938E23A03C58a4e93A66AAEC20, maxFeePerGas: 20000000000 baseFee: 186671733869 (supplied gas 8000000)"
```

**Comment by imagobea** [comment]
Hopefully, continues in #1262
